edx_exams/apps/api/v1/views.py

```diff
@@ -1 +1,175 @@
+from rest_framework import status
+from rest_framework.response import Response
+from rest_framework.views import APIView
+
+from edx_exams.apps.api.permissions import StaffUserPermissions
+from edx_exams.apps.api.serializers import ExamSerializer
+from edx_exams.apps.core.exam_types import get_exam_type
+from edx_exams.apps.core.models import CourseExamConfiguration, Exam
+
+log = logging.getLogger(__name__)
+
+
+class CourseExamsView(APIView):
+    """
+    View to modify exams for a specific course.
+
+    Given a list of exam data for a course, this view will either create a new exam (if one doesn't exist), or modify
+    an existing exam. Any course exams missing from the request will be marked as inactive.
+
+    HTTP PATCH
+    Creates a new Exam.
+    Expected PATCH data: [{
+        "content_id": 123,
+        "exam_name": "Test Examination",
+        "time_limit_mins": 30,
+        "due_date": "2021-08-04",
+        "exam_type": "proctored",
+        "hide_after_due": False,
+        "is_active": True
+    }]
+    **PATCH data Parameters**
+        * content_id: This will be the pointer to the id of the piece of course_ware which is the proctored exam.
+        * exam_name: This is the display name of the Exam (Midterm etc).
+        * time_limit_mins: Time limit (in minutes) that a student can finish this exam.
+        * due_date: The date on which the exam is due
+        * exam_type: Type of exam, i.e. timed, proctored
+        * hide_after_due: Whether the exam will be hidden from the learner after the due date has passed.
+        * is_active: Whether this exam will be active.
+    **Exceptions**
+        * HTTP_400_BAD_REQUEST
+    """
+
+    authentication_classes = (JwtAuthentication,)
+    permission_classes = (StaffUserPermissions,)
+
+    @classmethod
+    def update_exam(cls, exam_object, fields):
+        """
+        Given an exam object, update to the given fields value
+        """
+        for attr, value in fields.items():
+            setattr(exam_object, attr, value)
+        exam_object.save()
+
+        log.info(
+            "Updated existing exam=%(exam_id)s",
+            {
+                'exam_id': exam_object.id,
+            }
+        )
+
+    @classmethod
+    def create_exam(cls, fields):
+        """
+        Create a new exam based on the given fields
+        """
+        exam = Exam.objects.create(resource_id=str(uuid.uuid4()), **fields)
+
+        log.info(
+            "Created new exam=%(exam_id)s",
+            {
+                'exam_id': exam.id,
+            }
+        )
+
+    @classmethod
+    def handle_exams(cls, request_exams_list, course_exams_qs, course_id):
+        """
+        Decide how exams should be updated or created
+        """
+
+        for exam in request_exams_list:
+            # should only be one object per exam type per content_id
+            type_exam_qs = course_exams_qs.filter(content_id=exam['content_id'], exam_type=exam['exam_type'])
```

Comment:
is this going to do one SQL query per exam in the request list? maybe it would make sense to get all the course exams into a dict by content ID at the top of the loop and use that

Reply:
Created dict to organize info 👍